### PR TITLE
[New] Query feature count and extent

### PR DESCRIPTION
**Pull request by fcamas**
## Description

This PR implements `Query feature count and extent` in `Analysis` category.
URL to [README](https://devtopia.esri.com/runtime/common-samples/tree/main/designs/QueryFeatureCountAndExtent)

## Linked Issue(s)

- `swift/issues/`

## How To Test

Use the button select state then use the picker to zoom to the extent of the state specified. Use the button to count the features in the current extent.

## Screenshots

Query-feature-count-and-extent-sample

### CODE REVIEW

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    @State private var showFeatureCountBar = false
+    @State private var showAlert = false
+    @State private var alertMessage = ""
+    
+    var body: some View {
+        ZStack {
+            VStack {
+                MapViewReader { proxy in
+                    MapView(map: model.map)
+                    // Perform update viewpoint.
```
[the remark this answers is not in the document]
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift]: Yes, thank you for the input.

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    var body: some View {
+        VStack {
+            MapViewReader { proxy in
+                MapView(map: model.map)
+                // Perform update viewpoint.
+                    .onViewpointChanged(kind: .boundingGeometry) { newViewpoint in
+                        model.viewpoint = newViewpoint
+                    }
+                // Perform query when the selected state changes.
+                    .onChange(of: model.selectedState) { _ in
```
[the remark this answers is not in the document]
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift]: Thank you for clarifying. I really appreciate it.

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        init() {
+            featureTable = ServiceFeatureTable(url: .medicareHospitalSpendLayer)
+            featureLayer = FeatureLayer(featureTable: featureTable)
+            map.addOperationalLayer(featureLayer)
+        }
+        
+        /// Performs a query based on the selected state and updates the map viewpoint to encompass the results.
+        /// - Parameter mapViewProxy: The proxy to update the map viewpoint.
+        /// - Throws: An error if the query fails.
+        func performQuery(on mapViewProxy: MapViewProxy?) async throws {
```
[the remark this answers is not in the document]
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift]: Thanks for the suggestion! I’ll update it to use queryExtent(stateAbbreviation:) and make selectedState a state variable.

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                    await mapViewProxy?.setViewpointGeometry(combinedExtent)
+                }
+            }
+        }
+        
+        /// Counts the number of features within the specified viewpoint extent.
+        /// - Parameter withinViewpoint: The viewpoint defining the extent to query.
+        /// - Returns: The count of features within the specified extent.
+        /// - Throws: An error if the count operation fails.
+        func performCountOnVisibleExtent(withinViewpoint: Viewpoint?) async throws {
```
[the remark this answers is not in the document]
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift]: Thanks for the suggestion! I’ll change it to countWithinExtent(extent:)

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+            "WV", "WI", "WY"
+        ]
+        
+        init() {
+            featureTable = ServiceFeatureTable(url: .medicareHospitalSpendLayer)
+            featureLayer = FeatureLayer(featureTable: featureTable)
+            map.addOperationalLayer(featureLayer)
+        }
+        
+        /// Queries the extent for the selected state.
```
Comment (CalebRas) [inline Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift:133]:
```suggestion
        /// Queries the extent for the specified state.
```

File: Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                    // Perform query and update the viewpoint when the selected state changes.
+                    guard let state = selectedState else { return }
+                    do {
+                        if let combinedExtent = try await model.queryExtent(stateAbbreviation: state) {
+                            await proxy.setViewpointGeometry(combinedExtent)
+                            showFeatureCountBar = false
+                        }
+                    } catch {
+                        self.error = error
+                    }
```
Comment (CalebRas) [inline Shared/Samples/Query feature count and extent/QueryFeatureCountAndExtentView.swift:54]:
```suggestion
                    }
                    selectedState = nil
```

This allows the same state to be selected twice in a row.

File: Shared/Samples/Query feature count and extent/query-feature-count-and-extent.png
Comment (CalebRas) [inline Shared/Samples/Query feature count and extent/query-feature-count-and-extent.png:1]: Could you update the screenshot with the menu closed to show case the map?

**Review by CalebRas — APPROVED** [review]
Looks good! Just a few minor things. Thank you for making these changes Fredy

File: Shared/Samples/Query feature count and extent/README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+## How to use the sample
+
+Use the button select a state, then use the picker to zoom to the extent of the state specified. Use the button to count the features in the current extent.
+
+## How it works
+
+Querying by state abbreviation:
+
+1. A `QueryParameters` object is created with a `WhereClause`.
+2. `FeatureTable.QueryExtentAsync` is called with the `QueryParameters` object to obtain the extent that contains all matching features.
```
Comment (pgruenler) [inline Shared/Samples/Query feature count and extent/README.md:20]:
```suggestion
2. `FeatureTable.QueryFeatures` is called with the `QueryParameters` object to obtain the extent that contains all matching features.
```

File: Shared/Samples/Query feature count and extent/README.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+3. The extent is converted to a `Viewpoint`, which is passed to `MapView.SetViewpointAsync`.
+
+Counting features in the current extent:
+
+1. The current visible extent is obtained from a call to `MapView.onViewpointChanged(kind:)`.
+2. A `QueryParameters` object is created with the visible extent and a defined `SpatialRelationship` (in this case 'intersects').
+3. The count of matching features is obtained from a call to `FeatureTable.QueryFeatureCountAsync`.
```
Comment (pgruenler) [inline Shared/Samples/Query feature count and extent/README.md:27]:
```suggestion
3. The count of matching features is obtained from a call to `FeatureTable.QueryFeatureCount`.
```

File: Shared/Samples/Query feature count and extent/README.metadata.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    "keywords": [
+        "count",
+        "feature layer",
+        "feature table",
+        "features",
+        "filter",
+        "number",
+        "query",
+        "FeatureTable.QueryExtentAsync",
+        "FeatureTable.QueryFeatureCountAsync",
```
Comment (pgruenler) [inline Shared/Samples/Query feature count and extent/README.metadata.json:17]:
```suggestion
        "FeatureTable.QueryExtent",
        "FeatureTable.QueryFeatureCount",
```

File: Shared/Samples/Query feature count and extent/README.metadata.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        "MapView",
+        "QueryParameters",
+        "QueryParameters.Geometry",
+        "QueryParameters.SpatialRelationship",
+        "QueryParameters.WhereClause"
+    ],
+    "redirect_from": [],
+    "relevant_apis": [
+        "FeatureTable.QueryExtentAsync",
+        "FeatureTable.QueryFeatureCountAsync",
```
Comment (pgruenler) [inline Shared/Samples/Query feature count and extent/README.metadata.json:27]:
```suggestion
        "FeatureTable.QueryExtent",
        "FeatureTable.QueryFeatureCount",
```

File: Shared/Samples/Query feature count and extent/README.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+![Image of Query feature count and extent sample](query-feature-count-and-extent.png)
+
+## Use case
+
+Queries can be used to search for features in a feature table using text entry. This is helpful for finding a specific feature by name in a large feature table. A query can also be used to count features in an extent. This could be used to count the number of traffic incidents in a particular region when working with an incident dataset for a larger area.
+
+## How to use the sample
+
+Use the button select a state, then use the picker to zoom to the extent of the state specified. Use the button to count the features in the current extent.
```
Comment (pgruenler) [inline Shared/Samples/Query feature count and extent/README.md:13]:
```suggestion
Use the button Select State, then use the picker to zoom to the extent of the state specified. Use the button to count the features in the current extent.
```
I think capitalizing "Select State" makes it clearer that it is referring to the button as it is in the UI.

**Review by pgruenler — COMMENTED** [review]
Code looks good. Just a few comments.

File: Shared/Samples/Query feature count and extent/query-feature-count-and-extent.png
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/query-feature-count-and-extent.png:1]: Thank you for your feedback.

File: Shared/Samples/Query feature count and extent/README.md (hunk 11 above)
Comment (fcamas) [reply Shared/Samples/Query feature count and extent/README.md]: I appreciate the feedback; thank you!

**Review by CalebRas — APPROVED** [review]

**Review by pgruenler — APPROVED** [review]

**Pull request merged**